### DISCUSSION TITLE
이 라이브러리 사용 목적에 관하여

### DISCUSSION THREAD

**Discussion by tedKim5178** [skydoves/compose-stable-marker#7] [Q&A] [answered]
안녕하세요. 
이 라이브러리는 stable, immutable 같은 어노테이션을 사용하고 싶지만, compose-runtime 라이브러리는 사용하고 싶지 않을때 사용하는 건가요?
궁극적으론 용량 감소 목적이라고 봐도 괜찮은가요? (compose-runtime에는 해당 어노테이션 말고 다른것들도 많기때문에)

**Comment by skydoves** [comment, marked as answer]
안녕하세요,

`compose-stable-marker`는 data layer와 같이 Android 및 Compose와 관련한 의존성으로부터 자유롭고 싶은 모듈(JVM 등)에서 Compose runtime에 대한 의존성을 갖지 않고, `Stable` 및 `Immutable`을 사용하기 위함입니다. 

아쉽게도 Compose에서는 멀티 모듈 관계의 경우 data 모델이 모두 불가변의 property 들로 구성이 되어있다고 하더라도, Compose compiler에 의해 unstable로 해석됩니다. 

단순히 data layer 모듈에서 이를 해결하기 위해 [compose-runime](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/runtime/runtime/src/commonMain/kotlin/androidx/compose/runtime/) package를 통째로 포함시키는 것은 대부분의 class들이 runtime 내부적으로만 사용된다고 하더라도, 불필요한 의존성을 부여하기 때문에 이로부터 조금 더 자유로운 유즈케이스에 활용되기를 기대하며 만들어진 라이브러리입니다.

말씀하신 것처럼 용량 감소뿐만 아니라, 개발 시 의도적이든 의도적이지 않던 `compose-runtime`에 대한 **의존성을 가질 수 있는 가능성 자체를 배제하기 위함**입니다. `@Composable`를 포함한 각종 런타임 어노테이션, `SideEffect`, `LaunchedEffect`, `DisposableEffect`, `snapshotFlow`, `StateFlow<T>.collectAsState()` 등 여전히 많은 부분들이 외부로 노출되어 있습니다. 

특히 Android나 Compose에 대한 의존성이 완전 무결한 core-level의 라이브러리 및 SDK를 개발해야 하는 경우 조금 더 의미 있게 사용될 수 있다고 보실 수 있습니다. 가령, [Stream Chat Android SDK의 core 패키지에 사용되는 모델](https://github.com/GetStream/stream-chat-android/tree/3a352efa1477332e206b4de0aabff7e3f84a3318/stream-chat-android-core/src/main/java/io/getstream/chat/android/models)의 경우 Compose를 제외한 XML 및 다른 package에서도 상호 참조가 가능해야 하는데, 이때 KMP로 구현된 `compose-runtime` 의존성으로 인해 core에서 `compose-runtime`에 대한 참조 가능성이 생기는 사실 자체를 완전히 배제하기 위함입니다.

Compose Runtime은 Compose Compiler와 독립적이기 때문에 Compose Compiler가 사용되지 않는 모듈로서의 참조는 언제나 기대하는 동작을 가져오지만, Compose Compiler가 사용되는 모듈에서 문제가 발생할 수 있습니다. (가령, `@ReadOnlyComposable`이나 `@NonRestartableComposable` 등이 오용된 경우)

도움이 되셨기를 바랍니다!

**Reply by tedKim5178** [reply]
의존성 분리가 메인 목적이고 용량은 부가적인 효과라고 생각하면 되겠네요. 답변 감사합니다. 큰 도움 되었습니다!!!
